File: tests/drift_kinetic/drift_kinetic_curv_gradb_ex4.py
#!/usr/bin/env python3

import json
import os
import sys

import numpy as np

sys.path.append(os.path.join(os.path.dirname(__file__), "../../tools"))

from lib.plot_utils import figure, subplot

HERE = os.path.dirname(os.path.abspath(__file__))
OUT = os.path.join(HERE, "output", "drift_kinetic_curv_gradb_ex4")
DK_DIR = os.path.join(OUT, "drift_kinetic")
KIN_DIR = os.path.join(OUT, "eccapfim")

COLOR_KIN = "red"
COLOR_DK = "blue"
COLOR_TH = "black"

LABELSIZE = 17
TICKSIZE = 13
LEGENDSIZE = 12
PANELSIZE = 17
TITLESIZE = 16
OFFSETSIZE = 15


def load_config(path):
    with open(os.path.join(path, "config.json")) as f:
        return json.load(f)


def azimuthal_center(config):
    # Returns the axis location of the SetAzimuthalField magnetic preset.
    for preset in config["Presets"]:
        if preset.get("command") != "SetMagneticField":
            continue
        setter = preset["setter"]
        if setter.get("name") == "SetAzimuthalField":
            return np.array(setter["center"], dtype=float)
    raise KeyError("SetAzimuthalField preset is not found in the config")


def load_trace(path):
    return np.loadtxt(os.path.join(path, "temporal", "particle_trace.txt"), skiprows=1)


def mark_ends(ax, xs, ys, color):
    # Hollow dot at the start of the trajectory, solid dot at the end.
    ax.plot(xs[0], ys[0], "o", mfc="white", mec=color, mew=1.5, ms=7, zorder=5)
    ax.plot(xs[-1], ys[-1], "o", color=color, ms=7, zorder=5)


def style_axes(ax, panel):
    ax.minorticks_on()
    # Ticks on every side, pointing inward; x labels only on the bottom.
    ax.tick_params(axis="both", which="both", direction="in",
        top=True, bottom=True, left=True, right=True, labelsize=TICKSIZE)
    ax.tick_params(axis="x", which="both", labelbottom=True, labeltop=False)
    ax.tick_params(axis="y", which="both", labelleft=True, labelright=False)
    ax.text(0.88, 0.98, panel, transform=ax.transAxes,
        ha="right", va="top", fontsize=PANELSIZE,
        bbox=dict(facecolor="white", edgecolor="none", alpha=0.6,
            boxstyle="round,pad=0.2"))


def main():
    cfg_dk = load_config(DK_DIR)
    cfg_kin = load_config(KIN_DIR)

    q = float(cfg_dk["Particles"][0]["q"])
    m = float(cfg_dk["Particles"][0]["m"])

    center = azimuthal_center(cfg_dk)
    cx, cy = center[0], center[1]

    dk = load_trace(DK_DIR)
    kin = load_trace(KIN_DIR)

    t_dk, x_dk, y_dk, z_dk = dk[:, 0], dk[:, 1], dk[:, 2], dk[:, 3]
    p_par_dk, p_perp_dk, mu_dk = dk[:, 4], dk[:, 5], dk[:, 6]
    t_kin, x_kin, y_kin, z_kin = kin[:, 0], kin[:, 1], kin[:, 2], kin[:, 3]

    # Field magnitude at the guiding center, from the conserved invariants
    # (B = m p_perp^2 / 2mu) so it matches what the pusher interpolated.
    b_gc = m * p_perp_dk[0] ** 2 / (2.0 * mu_dk[0])
    omega_e = abs(q) * b_gc / m
    rho = p_perp_dk[0] / (abs(q) * b_gc)

    # Field-line radius = guiding-center distance from the axis; parallel and
    # perpendicular velocities from the invariants. The rigid-rotor field has
    # both drifts along z; the total is what should be subtracted:
    #   V_kappa = + m v_par^2  / (q |B| R)    (curvature)
    #   V_gradB = - m v_perp^2 / (2 q |B| R)  (grad-B, opposes curvature here)
    R = np.hypot(x_dk[0] - cx, y_dk[0] - cy)
    v_par = p_par_dk[0] / m
    v_perp = p_perp_dk[0] / m
    v_kappa = m * v_par ** 2 / (q * b_gc * R)
    v_gradb = -m * v_perp ** 2 / (2.0 * q * b_gc * R)
    v_drift = v_kappa + v_gradb
    z_gc = z_dk[0]

    # Steps are stored in 1/w_pe; Omega_e = |q| B0 / m = 1 by design, so the
    # step in units of 1/Omega_e is just the configured dt (use it directly to
    # keep exact 0.1 / 10 instead of interpolation-noise digits from b_gc).
    tau_kin = float(cfg_kin["Geometry"]["dt"])
    tau_dk = float(cfg_dk["Geometry"]["dt"])
    label_kin = rf"kinetic: $\tau = {tau_kin:g}/\Omega_e$"
    label_dk = rf"drift-kinetic: $\tau = {tau_dk:g}/\Omega_e$"

    fig, gs = figure(ncols=3, nrows=1, figsize=(14, 4.5))

    # (a) Trajectory projected on the (z, x) plane: the cyclotron orbit lies in
    # this plane (B ~ e_y at the seed), so it shows as a circle of radius rho
    # that the curvature + grad-B drift shifts along z.
    ax = subplot(fig, gs, 0, 0)
    ax.plot(z_kin, x_kin, color=COLOR_KIN, lw=1.0, label=label_kin)
    ax.plot(z_dk, x_dk, color=COLOR_DK, lw=2.0, label=label_dk)
    mark_ends(ax, z_kin, x_kin, COLOR_KIN)
    mark_ends(ax, z_dk, x_dk, COLOR_DK)
    ax.set_xlabel(r"$z,\ c/\omega_{pe}$", fontsize=LABELSIZE)
    ax.set_ylabel(r"$x,\ c/\omega_{pe}$", fontsize=LABELSIZE)
    # Frame on the actual particle trajectory, not the whole model domain;
    # use an equal span on both axes so the panel stays square.
    hs = np.concatenate([z_kin, z_dk])
    vs = np.concatenate([x_kin, x_dk])
    pad = 0.3
    hc = 0.5 * (hs.min() + hs.max())
    vc = 0.5 * (vs.min() + vs.max())
    half = 0.5 * max(hs.max() - hs.min(), vs.max() - vs.min()) + pad
    ax.set_xlim(hc - half, hc + half)
    ax.set_ylim(vc - half, vc + half)
    ax.set_aspect("equal", adjustable="box")
    ax.text(0.03, 0.22, r"curv. + $\nabla B$ drift",
        transform=ax.transAxes, ha="left", va="bottom", fontsize=TITLESIZE)
    ax.legend(loc="lower left", fontsize=LEGENDSIZE)
    style_axes(ax, "(a)")

    # (b) Drift-subtracted z(t) - (V_gradB + V_kappa) t; gyration within +-rho_e.
    ax = subplot(fig, gs, 1, 0)
    z_kin_dev = z_kin - v_drift * t_kin
    z_dk_dev = z_dk - v_drift * t_dk
    ax.plot(t_kin, z_kin_dev, color=COLOR_KIN, lw=1.0, label=label_kin)
    ax.plot(t_dk, z_dk_dev, color=COLOR_DK, lw=2.0, label=label_dk)
    ax.axhline(z_gc + rho, color="gray", ls=":", lw=1.0, label=r"theory $\pm\ \rho_e$")
    ax.axhline(z_gc - rho, color="gray", ls=":", lw=1.0)
    mark_ends(ax, t_kin, z_kin_dev, COLOR_KIN)
    mark_ends(ax, t_dk, z_dk_dev, COLOR_DK)
    ax.set_ylim(z_gc - 1.5 * rho, z_gc + 2.5 * rho)
    ax.set_xlabel(r"$t,\ \omega_{pe}^{-1}$", fontsize=LABELSIZE)
    ax.set_ylabel(r"$z(t) - (V_{\nabla B} + V_\kappa)\,t,\ c/\omega_{pe}$", fontsize=LABELSIZE)
    ax.legend(loc="upper left", fontsize=LEGENDSIZE)
    style_axes(ax, "(b)")

    # (c) Relative error of the drift-kinetic z-velocity vs the total drift.
    ax = subplot(fig, gs, 2, 0)
    v_dk = np.gradient(z_dk, t_dk)
    rel_err = (v_dk - v_drift) / v_drift
    ax.plot(t_dk, rel_err, color=COLOR_DK, lw=1.5, label=label_dk)
    ax.axhline(0.0, color=COLOR_TH, ls="--", lw=1.0)
    ymax = np.max(np.abs(rel_err))
    if ymax > 0:
        ax.set_ylim(-1.1 * ymax, 1.1 * ymax)
    ax.set_xlabel(r"$t,\ \omega_{pe}^{-1}$", fontsize=LABELSIZE)
    ax.set_ylabel(r"$(V(t) - V_{tot}) / V_{tot}$", fontsize=LABELSIZE)
    ax.ticklabel_format(axis="y", style="sci", scilimits=(0, 0))
    ax.yaxis.get_offset_text().set_fontsize(OFFSETSIZE)
    ax.legend(loc="best", fontsize=LEGENDSIZE)
    style_axes(ax, "(c)")

    print(f"|B| at guiding center  = {b_gc:.6e}")
    print(f"V_kappa (curvature)     = {v_kappa:.6e}")
    print(f"V_gradB (gradient)      = {v_gradb:.6e}")
    print(f"V_tot   (theory)        = {v_drift:.6e}")
    print(f"V_tot   (drift-kinetic) = {np.mean(v_dk):.6e}")
    print(f"max relative error      = {np.max(np.abs(rel_err)):.3e}")

    fig.tight_layout(pad=1.2, w_pad=2.5)
    image = os.path.join(OUT, "drift_kinetic_curv_gradb_ex4.png")
    fig.savefig(image, dpi=150)
    print(f"Saved figure to {image}")


if __name__ == "__main__":
    main()

File: tests/drift_kinetic/drift_kinetic_curv_gradb_ex4.cpp
#include "src/impls/drift_kinetic/simulation.h"
#include "src/impls/eccapfim/simulation.h"
#include "src/utils/configuration.h"
#include "tests/common.h"

/**
 * @brief Combined curvature + grad-B drift comparison: drift_kinetic vs eccapfim.
 *
 * A prescribed "rigid-rotor" azimuthal field B = Omega * (-(y - c_y), (x - c_x),
 * 0), so the field lines are circles around the z-axis and the magnitude grows
 * linearly with the radius, |B| = Omega * r. Unlike the constant-|B| case
 * (curv_ex3), here grad|B| = Omega e_r != 0, so BOTH drifts are present and add
 * along z:
 *   V_kappa  = + m v_par^2  / (q |B| R) z   (curvature, from parallel streaming)
 *   V_gradB  = - m v_perp^2 / (2 q |B| R) z (grad-B, opposes curvature here)
 * with R the field-line radius = guiding-center distance from the axis and
 * |B| = Omega R there. The setter is SetAzimuthalField with a finite `radius`,
 * which gives the linear field B = (value / radius) (-r_y, r_x, 0); choosing
 * value = B0 and radius = R makes |B|(R) = B0. A single electron is seeded at
 * the same coordinate for both runs (transferred to the guiding center
 * automatically for drift_kinetic). Only a quarter of the field-line circle is
 * covered: the axis sits one margin in from the lower-left transverse corner and
 * the box spans the first quadrant (R + 2*margin) so the whole arc + gyration
 * fit. Each run writes its per-step coordinate trace into its own subdirectory:
 * <out>/{drift_kinetic,eccapfim}/temporal/particle_trace.txt.
 */

static constexpr char help[] =
  "Curvature + grad-B drift comparison: drift_kinetic vs eccapfim.\n";

static constexpr PetscReal cell = 0.125;
static constexpr PetscReal B0_value = 1.0;  // |B| at the guiding center (r = R)
static constexpr PetscReal R_value = 15.0;  // field-line radius; |B| = (B0/R) r
static constexpr PetscReal margin = 1;  // gyro-radius + interpolation stencil around the arc

// Only a quarter of the field-line circle is covered: the axis (|| z) sits one
// margin in from the lower-left transverse corner, and the box spans the first
// quadrant (R + 2*margin) so the whole quarter arc + gyration fit inside.
static constexpr PetscInt grid_nx =
  static_cast<PetscInt>((R_value + 2 * margin) / cell);
static constexpr PetscInt grid_ny = grid_nx;
static constexpr PetscInt grid_nz = 8;

static constexpr PetscReal total_time = 300.0;
static constexpr PetscReal dt_kinetic = 0.1;
static constexpr PetscReal dt_drift = 10.0;

static const Vector3R center(margin, margin, 0.0);
// Particle on the +x side of the circle: there e_phi = +y, so p0 = (v_perp, v_par, 0).
static const Vector3R r0(center[X] + R_value, center[Y], grid_nz * cell / 2);
static const Vector3R p0(0.08, 0.08, 0.0);

void overwrite_config(
  const std::string& simulation, const std::string& out_dir, PetscReal step);

int main(int argc, char** argv)
{
  PetscFunctionBeginUser;
  PetscCall(PetscInitialize(&argc, &argv, nullptr, help));

  const std::filesystem::path base = get_out_dir(__FILE__);
  const std::string dk_dir = (base / "drift_kinetic").string();
  const std::string kin_dir = (base / "eccapfim").string();

  overwrite_config("drift_kinetic", dk_dir, dt_drift);
  std::filesystem::create_directories(dk_dir);
  Configuration::save(dk_dir);
  {
    drift_kinetic::Simulation simulation;
    PetscCall(simulation.initialize());
    PetscCall(simulation.calculate());
    PetscCall(simulation.finalize());
  }

  overwrite_config("eccapfim", kin_dir, dt_kinetic);
  std::filesystem::create_directories(kin_dir);
  Configuration::save(kin_dir);
  {
    eccapfim::Simulation simulation;
    PetscCall(simulation.initialize());
    PetscCall(simulation.calculate());
    PetscCall(simulation.finalize());
  }

  PetscCall(PetscFinalize());
  PetscFunctionReturn(PETSC_SUCCESS);
}

void overwrite_config(
  const std::string& simulation, const std::string& out_dir, PetscReal step)
{
  dx = dy = dz = cell;
  geom_nx = grid_nx;
  geom_ny = grid_ny;
  geom_nz = grid_nz;
  geom_x = geom_nx * dx;
  geom_y = geom_ny * dy;
  geom_z = geom_nz * dz;

  dt = step;
  geom_nt = static_cast<PetscInt>(std::llround(total_time / dt));
  geom_t = geom_nt * dt;

  Configuration::overwrite({
    {"Simulation", simulation},
    {"OutputDirectory", out_dir},
    {
      "Geometry",
      {
        {"x", geom_x},
        {"y", geom_y},
        {"z", geom_z},
        {"t", geom_t},
        {"dx", dx},
        {"dy", dy},
        {"dz", dz},
        {"dt", dt},
        {"diagnose_period", dt},
        {"da_boundary_x", "DM_BOUNDARY_PERIODIC"},
        {"da_boundary_y", "DM_BOUNDARY_PERIODIC"},
        {"da_boundary_z", "DM_BOUNDARY_PERIODIC"},
      },
    },
    {
      "Particles",
      {{
        {"sort_name", "electrons"},
        {"Np", 1},
        {"n", +1.0},
        {"q", -1.0},
        {"m", +1.0},
        {"T", +0.1},
      }},
    },
    {
      "Presets",
      {
        {
          {"command", "SetMagneticField"},
          {"field", "B0"},
          {"field_axpy", "B"},
          {"setter", {
            {"name", "SetAzimuthalField"},
            {"value", B0_value},
            {"center", {center[X], center[Y], 0.0}},
            {"radius", R_value},
          }},
        },
        {
          {"command", "SetParticles"},
          {"particles", "electrons"},
          {"coordinate", {{"name", "PreciseCoordinate"}, {"value", {r0[X], r0[Y], r0[Z]}}}},
          {"momentum", {{"name", "PreciseMomentum"}, {"value", {p0[X], p0[Y], p0[Z]}}}},
        },
      },
    },
  });
}
